In [191]:
import pretty_midi

In [241]:
midi_data = pretty_midi.PrettyMIDI('../songs/guren.mid')

In [193]:
print("duration:",midi_data.get_end_time())

duration: 100.31182558333343


In [194]:
for instrument in midi_data.instruments:
    print("instrument:", instrument.program);
    for note in instrument.notes:
        print(f'{note.pitch:10} {note.start:10} {note.end:10}')

instrument: 0
        61 2.655389041666667 2.8280411250000004
        61 2.8280411250000004   2.983428
        64 2.9930965166666668 3.146411566666667
        63 3.156770691666667 3.486190866666667
        59 3.486190866666667 3.6498650416666667
        59 3.6498650416666667  3.8038707
        57 3.813539216666667 4.144340608333334
        61 3.813539216666667 4.154009125
        61 4.154009125 4.3093960000000004
        64 4.319064516666667 4.472379566666667
        63 4.4827386916666665 4.812158866666667
        59 4.812158866666667   5.303872
        61 5.305253216666667 5.4799771250000004
        73 5.305253216666667 5.4799771250000004
        61 5.4799771250000004   5.635364
        73 5.4799771250000004 5.645032516666667
        64 5.645032516666667 5.798347566666667
        76 5.645032516666667 5.808016083333333
        63 5.808016083333333 6.138126866666667
        75 5.808016083333333 6.138126866666667
        59 6.138126866666667 6.301801041666667
        71 6.138126866666667

In [195]:
midi_data.instruments[0].notes

[Note(start=2.655389, end=2.828041, pitch=61, velocity=93),
 Note(start=2.828041, end=2.983428, pitch=61, velocity=88),
 Note(start=2.993097, end=3.146412, pitch=64, velocity=89),
 Note(start=3.156771, end=3.486191, pitch=63, velocity=88),
 Note(start=3.486191, end=3.649865, pitch=59, velocity=88),
 Note(start=3.649865, end=3.803871, pitch=59, velocity=89),
 Note(start=3.813539, end=4.144341, pitch=57, velocity=80),
 Note(start=3.813539, end=4.154009, pitch=61, velocity=90),
 Note(start=4.154009, end=4.309396, pitch=61, velocity=88),
 Note(start=4.319065, end=4.472380, pitch=64, velocity=89),
 Note(start=4.482739, end=4.812159, pitch=63, velocity=88),
 Note(start=4.812159, end=5.303872, pitch=59, velocity=88),
 Note(start=5.305253, end=5.479977, pitch=61, velocity=85),
 Note(start=5.305253, end=5.479977, pitch=73, velocity=95),
 Note(start=5.479977, end=5.635364, pitch=61, velocity=80),
 Note(start=5.479977, end=5.645033, pitch=73, velocity=94),
 Note(start=5.645033, end=5.798348, pitc

# IDeate

- sorted by start time
- notes can start in same time

# Problem

What note is playing in that time and group all note that play in same time, return notes as array and tell when that note start and end. (all notes in array should play full time)

หาว่า note อะไรที่เล่นอยู่ในช่วงเวลานั้นๆ
โดยให้ group ทุกโน็ตที่เล่นอยู่ในเวลาเดียวกันมาเป็น array และ บอกเวลาที่เล่น (ทุก note ใน array ต้องเล่นเต็มเวลา)

## Example

### notes
 - Note(start=2.655389, end=2.828041, pitch=61, velocity=93)
 - Note(start=2.828041, end=2.983428, pitch=61, velocity=88)
 - Note(start=2.993097, end=3.166412, pitch=64, velocity=89)
 - Note(start=3.156771, end=3.486191, pitch=63, velocity=88)
 - Note(start=3.156771, end=3.649865, pitch=59, velocity=88)

### answer
- start: 2.655389, end: 2.828041, notes: \[ 61 \]
- start: 2.828041, end: 2.983428, notes: \[ 61 \]
- start: 2.993097, end: 3.156771, notes: \[ 64 \]
- start: 3.156771, end: 3.166412, notes: \[ 64, 63, 59 \]
- start: 3.166412, end: 3.486191, notes: \[ 63, 59 \]
- start: 3.486191, end: 3.649865, notes: \[ 59 \]

In [196]:
notes = midi_data.instruments[0].notes[:6]

In [197]:
notes

[Note(start=2.655389, end=2.828041, pitch=61, velocity=93),
 Note(start=2.828041, end=2.983428, pitch=61, velocity=88),
 Note(start=2.993097, end=3.146412, pitch=64, velocity=89),
 Note(start=3.156771, end=3.486191, pitch=63, velocity=88),
 Note(start=3.486191, end=3.649865, pitch=59, velocity=88),
 Note(start=3.649865, end=3.803871, pitch=59, velocity=89)]

In [198]:
playing_notes = []

In [199]:
start_time = []
end_time = []

In [200]:
def find_min_start(notes):
    min_notes = []
    pop_index = []
    if len(notes) > 0:
        min_note = notes[0]
        min_start = min_note.start
        for i in range(len(notes)):
            note = notes[i]
            if note.start == min_start:
                min_notes.append(note)
                pop_index.append(i)
            else:
                break
        for i in range(len(pop_index)):
            index = pop_index[len(pop_index) - i - 1]
            notes.pop(index)
    return min_notes

In [201]:
min_notes = find_min_start(notes)

In [202]:
min_notes

[Note(start=2.655389, end=2.828041, pitch=61, velocity=93)]

In [203]:
notes

[Note(start=2.828041, end=2.983428, pitch=61, velocity=88),
 Note(start=2.993097, end=3.146412, pitch=64, velocity=89),
 Note(start=3.156771, end=3.486191, pitch=63, velocity=88),
 Note(start=3.486191, end=3.649865, pitch=59, velocity=88),
 Note(start=3.649865, end=3.803871, pitch=59, velocity=89)]

In [204]:
playing_notes.append(min_notes)
start_time.append(min_notes[0].start)

In [205]:
def find_first_end(notes):
    if len(notes) > 0:
        min_end = notes[0].end
        for note in notes:
            if note.end < min_end:
                min_end = note.end
        return min_end
    return -1

In [206]:
end = find_first_end(min_notes)

In [207]:
end

2.8280411250000004

In [208]:
def find_before(end, notes):
    if len(notes) > 0:
        min_note = notes[0]
        min_start = min_note.start
        if min_start > end:
            return []
        else:
            return find_min_start(notes)
    return []

In [209]:
notes_start_before_end = find_before(end, notes)

In [210]:
notes_start_before_end

[Note(start=2.828041, end=2.983428, pitch=61, velocity=88)]

In [211]:
notes

[Note(start=2.993097, end=3.146412, pitch=64, velocity=89),
 Note(start=3.156771, end=3.486191, pitch=63, velocity=88),
 Note(start=3.486191, end=3.649865, pitch=59, velocity=88),
 Note(start=3.649865, end=3.803871, pitch=59, velocity=89)]

In [212]:
end_time.append(notes_start_before_end[0].start)

In [213]:
current_time = end_time[-1]

In [214]:
current_time

2.8280411250000004

In [215]:
def check_next_playing_note(playing_notes, new_start_note, current_time):
    note_still_alive = []
    for note in playing_notes[-1]:
        if note.end > current_time:
            note_still_alive.append(note)
    for note in new_start_note:
        note_still_alive.append(note)
    return note_still_alive

In [216]:
check_next_playing_note(playing_notes, notes_start_before_end, current_time)

[Note(start=2.828041, end=2.983428, pitch=61, velocity=88)]

## Main Section

### notes
 - Note(start=2.655389, end=2.828041, pitch=61, velocity=93)
 - Note(start=2.828041, end=2.983428, pitch=61, velocity=88)
 - Note(start=2.993097, end=3.166412, pitch=64, velocity=89)
 - Note(start=3.156771, end=3.486191, pitch=63, velocity=88)
 - Note(start=3.156771, end=3.649865, pitch=59, velocity=88)

In [217]:
notes = midi_data.instruments[0].notes[:5]

In [218]:
notes[2].end = 3.166412

In [219]:
notes[3].start = 3.156771
notes[4].start = 3.156771

In [220]:
notes

[Note(start=2.655389, end=2.828041, pitch=61, velocity=93),
 Note(start=2.828041, end=2.983428, pitch=61, velocity=88),
 Note(start=2.993097, end=3.166412, pitch=64, velocity=89),
 Note(start=3.156771, end=3.486191, pitch=63, velocity=88),
 Note(start=3.156771, end=3.649865, pitch=59, velocity=88)]

In [231]:
def split_note_frame(notes):
    playing_notes = []
    start_time = []
    end_time = []
    notes_start_before_end = []
    current_notes = []
    current_time = 0
    
    while len(notes) > 0 or len(current_notes) > 0:
        
        if len(current_notes) < 1:
            # get current notes
            current_notes = find_min_start(notes)
            
        # add current notes to playing notes list
        playing_notes.append(current_notes)

        # set start time
        start_time.append(current_notes[0].start)

        # get first end in current playing
        end = find_first_end(current_notes)

        # find notes that start before playing notes has end
        notes_start_before_end = find_before(end, notes)


        # if have note start before end then cut as end_time
        if len(notes_start_before_end) > 0:
            end_time.append(notes_start_before_end[0].start)
            
        # if not have note start then it's end by time out
        else:
            end_time.append(end)

        # get current time to check time out
        current_time = end_time[-1]

        # get playing note, remove time out note and merge new note
        current_notes = check_next_playing_note(playing_notes, notes_start_before_end, current_time)

    return playing_notes, start_time, end_time

In [222]:
results, start , end = split_note_frame(notes)

running
all notes in song 5
dont have note to play
get current note [Note(start=2.655389, end=2.828041, pitch=61, velocity=93)]
have note start before current end [Note(start=2.828041, end=2.983428, pitch=61, velocity=88)]
have note start then end at 2.8280411250000004
now current time is 2.8280411250000004
now current notes is [Note(start=2.828041, end=2.983428, pitch=61, velocity=88)]
running
all notes in song 3
get current note [Note(start=2.828041, end=2.983428, pitch=61, velocity=88)]
have note start before current end []
don't have note start then end at 2.983428
now current time is 2.983428
now current notes is []
running
all notes in song 3
dont have note to play
get current note [Note(start=2.993097, end=3.166412, pitch=64, velocity=89)]
have note start before current end [Note(start=3.156771, end=3.486191, pitch=63, velocity=88), Note(start=3.156771, end=3.649865, pitch=59, velocity=88)]
have note start then end at 3.156771
now current time is 3.156771
now current notes is [N

### answer should be
- start: 2.655389, end: 2.828041, notes: \[ 61 \]
- start: 2.828041, end: 2.983428, notes: \[ 61 \]
- start: 2.993097, end: 3.156771, notes: \[ 64 \]
- start: 3.156771, end: 3.166412, notes: \[ 64, 63, 59 \]
- start: 3.166412, end: 3.486191, notes: \[ 63, 59 \]
- start: 3.486191, end: 3.649865, notes: \[ 59 \]

In [223]:
for i in range(len(results)):
    s = start[i]
    e = end[i]
    note = [ n.pitch for n in results[i] ]
    print("start:", s, "end:", e, "notes:", note)

start: 2.655389041666667 end: 2.8280411250000004 notes: [61]
start: 2.8280411250000004 end: 2.983428 notes: [61]
start: 2.9930965166666668 end: 3.156771 notes: [64]
start: 2.9930965166666668 end: 3.166412 notes: [64, 63, 59]
start: 3.156771 end: 3.486190866666667 notes: [63, 59]
start: 3.156771 end: 3.6498650416666667 notes: [59]


# Real case

In [242]:
notes = midi_data.instruments[0].notes

In [243]:
notes

[Note(start=2.655389, end=2.828041, pitch=61, velocity=93),
 Note(start=2.828041, end=2.983428, pitch=61, velocity=88),
 Note(start=2.993097, end=3.146412, pitch=64, velocity=89),
 Note(start=3.156771, end=3.486191, pitch=63, velocity=88),
 Note(start=3.486191, end=3.649865, pitch=59, velocity=88),
 Note(start=3.649865, end=3.803871, pitch=59, velocity=89),
 Note(start=3.813539, end=4.144341, pitch=57, velocity=80),
 Note(start=3.813539, end=4.154009, pitch=61, velocity=90),
 Note(start=4.154009, end=4.309396, pitch=61, velocity=88),
 Note(start=4.319065, end=4.472380, pitch=64, velocity=89),
 Note(start=4.482739, end=4.812159, pitch=63, velocity=88),
 Note(start=4.812159, end=5.303872, pitch=59, velocity=88),
 Note(start=5.305253, end=5.479977, pitch=61, velocity=85),
 Note(start=5.305253, end=5.479977, pitch=73, velocity=95),
 Note(start=5.479977, end=5.635364, pitch=61, velocity=80),
 Note(start=5.479977, end=5.645033, pitch=73, velocity=94),
 Note(start=5.645033, end=5.798348, pitc

In [244]:
results, start , end = split_note_frame(notes)

In [245]:
for i in range(len(results)):
    s = start[i]
    e = end[i]
    note = [ n.pitch for n in results[i] ]
    print("start:", s, "end:", e, "notes:", note)

start: 2.655389041666667 end: 2.8280411250000004 notes: [61]
start: 2.8280411250000004 end: 2.983428 notes: [61]
start: 2.9930965166666668 end: 3.146411566666667 notes: [64]
start: 3.156770691666667 end: 3.486190866666667 notes: [63]
start: 3.486190866666667 end: 3.6498650416666667 notes: [59]
start: 3.6498650416666667 end: 3.8038707 notes: [59]
start: 3.813539216666667 end: 4.144340608333334 notes: [57, 61]
start: 3.813539216666667 end: 4.154009125 notes: [61]
start: 4.154009125 end: 4.3093960000000004 notes: [61]
start: 4.319064516666667 end: 4.472379566666667 notes: [64]
start: 4.4827386916666665 end: 4.812158866666667 notes: [63]
start: 4.812158866666667 end: 5.303872 notes: [59]
start: 5.305253216666667 end: 5.4799771250000004 notes: [61, 73]
start: 5.4799771250000004 end: 5.635364 notes: [61, 73]
start: 5.4799771250000004 end: 5.645032516666667 notes: [73]
start: 5.645032516666667 end: 5.798347566666667 notes: [64, 76]
start: 5.645032516666667 end: 5.808016083333333 notes: [76]
s